pkg/controller/csi_handler.go

```diff
@@ -140,21 +144,50 @@ func (handler *csiHandler) deleteSnapshot(vsd *crdv1.VolumeSnapshotContent) erro
 	return nil
 }
 
-func (handler *csiHandler) ListSnapshots(vsd *crdv1.VolumeSnapshotContent) (*csi.SnapshotStatus, int64, error) {
-	if vsd.Spec.CSI == nil {
+func (handler *csiHandler) GetSnapshotStatus(content *crdv1.VolumeSnapshotContent) (*csi.SnapshotStatus, int64, error) {
+	if content.Spec.CSI == nil {
 		return nil, 0, fmt.Errorf("CSISnapshot not defined in spec")
 	}
 	ctx, cancel := context.WithTimeout(context.Background(), handler.timeout)
 	defer cancel()
 
-	csiSnapshotStatus, timestamp, err := handler.csiConnection.ListSnapshots(ctx, vsd.Spec.CSI.SnapshotHandle)
+	csiSnapshotStatus, timestamp, err := handler.csiConnection.GetSnapshotStatus(ctx, content.Spec.CSI.SnapshotHandle)
 	if err != nil {
-		return nil, 0, fmt.Errorf("failed to list snapshot data %s: %q", vsd.Name, err)
+		return nil, 0, fmt.Errorf("failed to list snapshot data %s: %q", content.Name, err)
 	}
 
 	return csiSnapshotStatus, timestamp, nil
 }
 
+func (handler *csiHandler) CheckandUpdateSnapshotStatusOperation(snapshot *crdv1.VolumeSnapshot, content *crdv1.VolumeSnapshotContent) (*crdv1.VolumeSnapshot, error) {
+	if snapshot.Status.AvailableAt != nil {
+		return handler.markSnapshotCompleted(snapshot)
+	}
+	status, _, err := handler.GetSnapshotStatus(content)
```

Comment:
The 2nd parameter int64 from GetSnapshotStatus is a timestamp.  It is the time when the snapshot is cut.  CSI plugin should set CreatedAt as a Unix nanoseconds time encoded as an int64.

GetSnapshotStatus(vsd *crdv1.VolumeSnapshotContent) (*csi.SnapshotStatus, int64, error)

From CSI spec:
  // Timestamp when the point-in-time snapshot is taken on the storage
  // system. The format of this field should be a Unix nanoseconds time
  // encoded as an int64. On Unix, the command `date +%s%N` returns the
  // current time in nanoseconds since 1970-01-01 00:00:00 UTC. This
  // field is REQUIRED.
  int64 created_at = 4;